python/lsst/daf/butler/core/connectionString.py

```diff
@@ -0,0 +1,168 @@
+    def dialect(self):
+        """Returns the dialect part of the connection string."""
+        return self.get_backend_name()
+
+    @property
+    def driver(self):
+        """Returns the driver part of the connection string."""
+        return self.get_driver_name()
+
+    def checkDialectDriverExist(self):
+        """Verifies that the targeted dialect and driver exist and are
+        installed.
+
+        Attempts to load the correct dialect and driver from SQLAlchemy. If
+        load is successfull, returns True, otherwise returns False.
+        """
+        try:
+            self._conStr.get_dialect()
+        except NoSuchModuleError:
+            return False
+        return True
+
+    def asDict(self):
+        """Returns connection string arguments as a dictionary."""
+        return self._conStr.translate_connect_args()
+
+
+class ConnectionStringFactory:
+    """Factory for `ConnectionString` instances.
+
+    Parameters
+    ----------
+    path : `str` or None, optional
+        Path to configuration file. Defaults to '~/.lsst/daf-auth.yaml'.
+    envVar : `str` or None, optional
+        Name of environment variable pointing to configuration file. Defaults
+        to "LSST_DB_AUTH"
+    authList : `list` [`dict`] or None, optional
+        Authentication configuration.
+
+    Notes
+    -----
+    At least one of ``path``, ``envVar``, or ``authList`` must be provided;
+    generally ``path`` should be provided as a default location.
+    """
+
+    keys = ('username', 'password', 'host', 'port', 'database')
+
+    def __init__(self, path=None, envVar=None, authList=None):
+        if path is None:
+            path = DB_AUTH_PATH
+        if envVar is None:
+            envVar = DB_AUTH_ENVVAR
+        self._dbAuth = DbAuth(path, envVar, authList)
+
+    def __str__(self):
+        return f"{self.__class__.__name__}@{self._dbAuth.authList}"
+
+    def fromConfig(self, config):
+        """Parses a Config 'db' string and other keywords and returns an
+        `sqlalchemy.engine.url.URL` object containing the components of a URL
+        used to connect to a database.
+
+        Parameters
+        ----------
+        config : `Config` or `RegistryConfig`
+            Config to parse for connection string components.
+        dbAuth : `DbAuth` or None
+            DbAuth instance pointing to the correct credentials file. If None
+            the default "~/.lsst/db-auth.yaml" or "LSST_DB_AUTH" env var will
+            be used.
+
+        Returns
+        -------
+        conStr : `sqlalchemy.engine.url.URL`
+            URL object suitable to be passed into `sqlalchemy.create_engine`
+            function.
+        """
+        regConf = RegistryConfig(config)
+        conStr = url.make_url(regConf['db'])
+
+        for key in self.keys:
+            if getattr(conStr, key) is None:
+                setattr(conStr, key, regConf.get(key))
+
+        # when the databsase is a file, host will be None and the path
```

Comment:
When the database is a file, there are no passwords, and you don't need to call DbAuth at all.

Reply:
When username or password are given in the `db` connection string or as explicit keys in the config the `DbAuth` is not instantiated. Is that ok?

Reply:
See below.